### PR TITLE
  166187865 favorite or unfavorite an article

### PR DESCRIPTION
**Pull request by alexxsanya**
#### What does this PR do?
- enable an authenticated user to favorite or unfavorite an article

#### Description of Task to be completed?

- add functionality to favorite an article
- add functionality to unfavorite an article
- add unit tests to the added functionalities

#### How should this be manually tested?
- fetch this repo locally in the terminal with `$ git fetch origin ft-favorite-articles-166187865`
- run `$ python manage.py migrate articles`
- Now test out the following.
   1. Favorite an article with this `POST ​/api/articles​/{slug}​/favorite​/` with no request body
   2. Unfavorite an article with `DELETE /api​/articles​/{slug}​/favorite​/`
   3. Retrieve all user's favorite articles with `GET /api/favorite-articles/`
   4. Do `GET /api/articles/{slug}/` for the favorited articles to confirm that the response body contains the `is_favorited` set to **True**
  5. Kindly confirm that the rest of the functionalities are still working well.

#### What are the relevant pivotal tracker stories?

- Story ID - [166187865](https://www.pivotaltracker.com/story/show/166187865)

### CODE REVIEW

**Review by Emmanuelaaron — CHANGES_REQUESTED** [review]
Great work @alexxsanya,
Only that a  user should get a valid response when they try to favorite/unfavorite an article that does not exist.
Please consider adding that validation.
other than that, great work is done

**Comment by alexxsanya** [comment]
> Great work @alexxsanya,
> Only that you need the user should get a valid response when they try to favorites/unfavorite an article that does not exist.
> Please consider adding that validation.
> other than that, great work done

**Comment by alexxsanya** [comment]
> Great work @alexxsanya,
> Only that a user should get a valid response when they try to favorite/unfavorite an article that does not exist.
> Please consider adding that validation.
> other than that, great work is done

@Emmanuelaaron noted. I am going to fix this. thanks

File: authors/apps/articles/views.py [hunk 1]
```diff
@@ -34,66 +40,203 @@ def get(self, request, **kwargs):
...
+        Favorite an article with specified slug
+        """
+        article = self.check_article_exists(**kwargs)
+        favorite = FavoriteAnArticle.objects.filter(
+            favorited_by=request.user, article=article
+        )
+        if not favorite:
+            serializer = self.serializer_class(data={"is_favorited": True})
+            serializer.is_valid(raise_exception=True)
+            serializer.save(article=article, favorited_by=request.user)
```
Comment (cdvx) [inline authors/apps/articles/views.py:155]: @alexxsanya , Please work on these codeclimate issues

File: authors/apps/articles/serializers.py [hunk 2]
```diff
@@ -26,4 +26,13 @@ class Meta:
...
-        
\ No newline at end of file
+
+class FavoriteAnArticleSerializer(serializers.ModelSerializer):
+
+    article = ArticleSerializer(required=False)
+    favorited_by = UserSerializer(required=False)
+
+    class Meta:
+        fields = '__all__'
```
Comment (jennizalwango) [inline authors/apps/articles/serializers.py:36]: @ alex good work but please help me understand why you would use `__all__` on line 36

File: authors/apps/articles/views.py (hunk 1 above)
Comment (alexxsanya) [reply authors/apps/articles/views.py:155]: @cdvx okay.

**Review by neelxie — COMMENTED** [review]
@alexxsanya have checked it out and getting an error on local but let me test it the review app.

**Review by neelxie — COMMENTED** [review]
@alexxsanya can this feature be tested with swagger because I am failing.

**Review by neelxie — APPROVED** [review]
Wonderful stuff @alexxsanya everything works flawlessly

**Review by HabibSentongo — APPROVED** [review]
@alexxsanya your work is good, and I have tested it and I'm confident it works fine. Kindly work on the few feedback from others so they may also approve this PR.

File: authors/apps/articles/views.py (hunk 1 above)
Comment (alexxsanya) [reply authors/apps/articles/views.py:155]: @cdvx this has been fixed

**Comment by alexxsanya** [comment]
> Great work @alexxsanya,
> Only that you need the user should get a valid response when they try to favorites/unfavorite an article that does not exist.
> Please consider adding that validation.
> other than that, great work done

@Emmanuelaaron this has been fixed

File: authors/apps/articles/serializers.py (hunk 2 above)
Comment (alexxsanya) [reply authors/apps/articles/serializers.py:36]: @jennizalwango using `__all__` to indicate that all fields in the model should be used.

**Review by Emmanuelaaron — APPROVED** [review]
having worked on all the feedback,
I believe we are good to go

**Review by jennizalwango — APPROVED** [review]

**Pull request merged**